clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+    green_array = np.squeeze(processed_granule['green'].values)
+    blue_array = np.squeeze(processed_granule['blue'].values)
+
+    red_array = np.nan_to_num(red_array, nan=0.0)
+    green_array = np.nan_to_num(green_array, nan=0.0)
+    blue_array = np.nan_to_num(blue_array, nan=0.0)
+
+    rgb_image = np.stack([red_array, green_array, blue_array], axis=-1)
+
+    if np.issubdtype(rgb_image.dtype, np.floating):
+        rgb_image = (rgb_image - np.min(rgb_image)) / (np.max(rgb_image) - np.min(rgb_image))
+    else:
+        raise
+
+    plt.imsave(f'{output_folder}/{processed_granule["date"]}.png', rgb_image)
+
+# %%
+output_folder = 'hvplot'
+if not os.path.exists(output_folder):
+    os.makedirs(output_folder)
+for processed_granule in processed_granules:
+    # Stack the RGB bands to create a multi-dimensional xarray DataArray
+    rgb_values = ['B04_Red', 'B03_Green', 'B02_Blue']
+    rgb_channel = xr.DataArray(rgb_values, dims='rgb_channel', name='rgb_channel')
+    rgb_stacked = xr.concat([
+        processed_granule['red'], 
+        processed_granule['green'], 
+        processed_granule['blue']
+    ], dim=rgb_channel)
+    # Ensure that the data type is appropriate for visualization
+    # Normalize if the data is in float format, or ensure it's within [0, 255] if it's integer
+    if np.issubdtype(rgb_stacked.dtype, np.floating):
+        rgb_stacked = (rgb_stacked - rgb_stacked.min()) / (rgb_stacked.max() - rgb_stacked.min()) * 255
+    rgb_stacked = rgb_stacked.astype(np.uint8)
+    title = f'Manuel Canal-Spanish Lake: {processed_granule["date"]}'
+    rgb_plot = rgb_stacked.hvplot.rgb(
+                        x='x', y='y',
+                        xlabel='Latitude', 
+                        ylabel='Longitude',
+                        bands='rgb_channel', 
+                        data_aspect=1,
+                        title=title,
+                        crs=ccrs.UTM(16))
+    hvplot.save(rgb_plot, f'{output_folder}/{processed_granule["date"]}.png', fmt='png')
+
+# %% [markdown]
+# ### Merge and Composite Data
+# 
+# You will notice for this watershed that:
+# 1. The raster data for each date are spread across 4 granules
+# 2. Any given image is incomplete because of clouds
+# 
+# **YOUR TASK:**
+# 1. For each band:
+#     1. For each date:
+#         1. Merge all 4 granules
+#         2. Mask any negative values created by interpolating from the nodata value of -9999 (`rioxarray` should account for this, but doesn't appear to when merging. If you leave these values in they will create problems down the line)
+#     2. Concatenate the merged DataArrays along a new date dimension
+#     3. Take the mean in the date dimension to create a composite image that fills cloud gaps
+#     4. Add the band as a dimension, and give the DataArray a name
+# 2. Concatenate along the band dimension
+
+# %%
+red_das, green_das, blue_das = [], [], []
+nir_das, swir1_das, swir2_das = [], [], []
+ignore_dates = ["2023-05-22",
```

Comment:
Without comments, I'm not really sure why you're ignoring these dates. If it's cloud cover, I'd recommend calculating that and do it programmatically -- it would make the code more reproducible in other locations/times.